### DISCUSSION TITLE
From a submodule, expose only certain options as flake outputs

### DISCUSSION THREAD

**Discussion by DrRuhe** [hercules-ci/flake-parts#187] [Q&A] [answered]
I declared a bunch of options for `perSystem.deployments.<name>`. These options get built into packages under `perSystem.deployments.<name>.public`. How do I expose this public option as a flake output without exposing the adjacent options?

I took a look at how mkTransposedPerSystemModule does it for the packages attribute. I fail to understand how the flake output actually gets populated, given that it defaults to  {} but seems to not be initialized from any `config.perSystem.XY`

**Comment by roberth** [comment, marked as answer]
I think you'd have to write the logic yourself, in order to expose only the `.public` value.
Maybe something like this?

```nix
flake.deployments =
  mapAttrs
    (system: v: mapAttrs (name: deployment: deployment.public) v.deployments)
    config.allSystems
```

Are you sure you want to have deployments for each of the `systems`? I would expect a flake attribute like `deployments` to pick a specific system type for me, similar to what we have with `nixosConfigurations`.

**Reply by roberth** [reply]
That is unless you want to use have a deployment script that can be launched from any of the `systems`, but affects a single deployment that is defined outside `perSystem`, in which case it does make sense.

Or maybe it's something else? What did you have in mind for `deployments`?

**Reply by DrRuhe** [reply]
I'm afraid I dont follow. With "pick a sytem type for me" do you mean a single output that does not contain several system attributes but rather the system gets chosen during evaluation?

By "deployment" I mean a collection of "services" where each service basically consists of a derivation to be run and some metadata like which arguments to pass etc.

Then I have several "renderers" which render the service definition into a format the "managers" can run. E.g. the "manager" honcho is running the output of the "procfile" renderer. 

The output from the renderers and managers is placed under `.public` which should be exposed.

**Reply by roberth** [reply]
I think I need better words for talking about system.
Anyway, considering that these are configurations for _local_ tools it makes sense to instantiate for each system, so ignore my question.

Was my suggestion with `flake.deployments` useful?

**Reply by DrRuhe** [reply]
Yeah, that worked. Thanks :)
